include/envoy/common/platform.h

```diff
@@ -0,0 +1,19 @@
+#pragma once
+
+// NOLINT(namespace-envoy)
+#if !defined(_MSC_VER)
+#define STACK_ALLOC_ARRAY(var, type, num) type var[num]
+
+#define PACKED_STRUCT(definition, ...) definition, ##__VA_ARGS__ __attribute__((packed))
+
+#else
+#include <malloc.h>
+
+#define STACK_ALLOC_ARRAY(var, type, num)                                                          \
+  type* var = static_cast<type*>(::_alloca(sizeof(type) * num))
```

Comment:
I just noticed this...does this actually work? What if 'type' has a constructor; it doesn't look like it would be called and you'd get uninitialized memory.

Reply:
you could loop over the array and used placed new, FWIW. You'd also need to take care of destructors.

Reply:
@jmarantz that's a good point -- the constructors/destructors definitely won't get called.

It works for now since the macro is only used with:
- `Buffer::RawSlice`
- `struct iovec`
- `uint64_t`
- `unsigned char`

and none of these types have constructors/destructors that do anything interesting. In the `Buffer::RawSlice`  +  `struct iovec` case, it looks like we always manually initialize the array (either through `getRawSlices` or looping over the array).

The best way to fix this probably depends on:
- how important it is that these arrays are on the stack
- does the solution need to be fully general or could we restrict it to the types above

Reply:
Great catch @jmarantz. @sesmith177 we definitely need to fix this somehow as it will produce a bunch of annoying and hard to find bugs. One option that comes to mind is having some wrapper RAII template object that does placement new and then would automatically do placement delete on scope exit?

Reply:
As an aside, could we just say that users have to compile Envoy with clang on windows? I'm wondering if that would also be a way to "fix" these problems?

Reply:
Another alternative here is to just edit the code-sites to just not use stack allocation. Is there a perf win from stack in this case? My thinking in general has been that if you want to have a lot of threads, then you don't want to allocate huge amounts of stack space to them, so you might not want to allocate potentially large things on the stack.

If that sounds right, I'd seal the deal by throwing a compile switch in clang to make that invalid, if such a switch exists (IIRC g++ has such a switch).

Reply:
A third option might be to use something like:

```
static_assert(std::is_trivially_destructible<type>::value, #type " must be trivially destructible");
```

in the macro/wrapper class to ensure that `type` has a trivial destructor and doesn't need to be deleted.

We're happy to go in any of the above directions.

It would be a significant effort to require clang on Windows. The tricky part would be ensuring that all third-party dependencies can be build with clang on Windows as well (i.e. do not require MSVC). When initially starting on this project, we made an attempt to compile everything with gcc/mingw, unfortunately there were dependencies that would not build on Windows with that toolchain.

Reply:
@mattklein123 @jmarantz We have a branch that replaces variable length arrays with `std::vector` (and adds `-Wvla` to ensure variable length arrays are invalid) here: https://github.com/greenhouse-org/envoy/commit/47c48a48f4ea71c00eb714674094bbd6d3d2a4c1

If it's decided that these arrays must be allocated on the stack, we will work on a wrapper template object and PR that instead

Reply:
I'm not in favor of switching to std::vector as I don't think we should take a perf hit due to a platform issue. If it's not too bad, I think I would go with the wrapper template. I don't think it will be that hard so can we try it first?

Reply:
Sounds good, we'll get that PR'd